"""–û—Ç—Ç–æ—á–∏—Ç—å –¥–∏–∑–∏–≥–Ω–µ"""

import sys
import os
import warnings
from pathlib import Path

warnings.filterwarnings(
    "ignore",
    category=DeprecationWarning,
    message=r".*sipPyTypeDict.*"
)

from PyQt6 import QtWidgets, uic, QtCore, QtGui
from PyQt6.QtCore import Qt, pyqtSignal, QSettings, QPoint
from PyQt6.QtWidgets import (
    QMainWindow, QWidget, QVBoxLayout, QHBoxLayout,
    QLabel, QComboBox, QToolButton, QTabWidget, QStatusBar,
    QMenu, QPushButton, QTextEdit, QLineEdit, QSplitter,
    QListWidget, QListWidgetItem
)
from PyQt6.QtGui import QIcon, QAction

try:
    from ..core.python.theme_util import SystemThemeDetector
    from ..core.python.theme_manager import setup_app_theme, ThemeManager
except ImportError as e:
    print(f"Import error: {e}")
    # Fallback: –¥–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –≤—Ä—É—á–Ω—É—é –¥–ª—è –æ—Ç–ª–∞–¥–∫–∏
    current_dir = Path(__file__).parent
    project_root = current_dir.parent.parent
    src_python_path = project_root / "worrrkspace" / "src" / "python"
    print(f"Trying to add path: {src_python_path}")
    sys.path.insert(0, str(src_python_path))
    from theme_util import SystemThemeDetector
    from theme_manager import setup_app_theme, ThemeManager

# –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º –≤–∏–¥–∂–µ—Ç—ã –∏–∑ –æ—Ç–¥–µ–ª—å–Ω—ã—Ö –º–æ–¥—É–ª–µ–π
try:
    from ..ui.widgets.base_widgets import ProfileDialog
    from ..ui.widgets.markdown_editor import MarkdownNoteTab
    from ..ui.widgets.table_editor import TableEditorTab
    from ..ui.widgets.graph_editor import GraphTab
    from ..ui.widgets.task_editor import TaskTab
except ImportError as e:
    print(f"Import error: {e}")
    # Fallback: –¥–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –≤—Ä—É—á–Ω—É—é –¥–ª—è –æ—Ç–ª–∞–¥–∫–∏
    current_dir = Path(__file__).parent
    project_root = current_dir.parent.parent
    widgets_python_path = project_root / "worrrkspace" / "ui" / "widgets"
    print(f"Trying to add path: {widgets_python_path}")
    sys.path.insert(0, str(widgets_python_path))
    from base_widgets import ProfileDialog
    from markdown_editor import MarkdownNoteTab
    from table_editor import TableEditorTab
    from graph_editor import GraphTab
    from task_editor import TaskTab

# –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º –ø–∞–Ω–µ–ª–∏ –∏–∑ –Ω–æ–≤–æ–π –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏–∏
try:
    from ..ui.panels import SolutionExplorer, ToolsPanel, ChatPanel, LogsPanel
except ImportError as e:
    print(f"Import error: {e}")
    # Fallback: –¥–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –≤—Ä—É—á–Ω—É—é –¥–ª—è –æ—Ç–ª–∞–¥–∫–∏
    current_dir = Path(__file__).parent
    project_root = current_dir.parent.parent
    panels_python_path = project_root / "worrrkspace" / "ui" / "panels"
    print(f"Trying to add path: {panels_python_path}")
    sys.path.insert(0, str(panels_python_path))
    from solution_explorer import SolutionExplorer
    from tools_panel import ToolsPanel
    from chat_panel import ChatPanel
    from logs_panel import LogsPanel
    


class DraggableTabBar(QtWidgets.QTabBar):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setMovable(True)  # –í–∫–ª—é—á–∞–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω–æ–µ –ø–µ—Ä–µ–º–µ—â–µ–Ω–∏–µ –¥–ª—è –õ–ö–ú
        self.setExpanding(False)

        # –ü–µ—Ä–µ–º–µ–Ω–Ω—ã–µ –¥–ª—è –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏—è –ü–ö–ú
        self.right_drag_start_position = None
        self.right_drag_current_position = None
        self.right_dragging = False
        self.right_drag_index = -1
        self.right_drag_pixmap = None

    def mousePressEvent(self, event):
        # –û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ–º –ü–ö–ú –¥–ª—è –º–µ–∂–æ–∫–æ–Ω–Ω–æ–≥–æ –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏—è
        if event.button() == Qt.MouseButton.RightButton:
            self.right_drag_start_position = event.position().toPoint()
            self.right_drag_current_position = self.right_drag_start_position
            self.right_dragging = True
            self.right_drag_index = self.tabAt(self.right_drag_start_position)

            # –°–æ–∑–¥–∞–µ–º pixmap –¥–ª—è –≤–∏–∑—É–∞–ª—å–Ω–æ–π –æ–±—Ä–∞—Ç–Ω–æ–π —Å–≤—è–∑–∏ –ø—Ä–∏ –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏–∏
            if self.right_drag_index >= 0:
                rect = self.tabRect(self.right_drag_index)
                self.right_drag_pixmap = self.grab(rect)

            # –ù–µ –≤—ã–∑—ã–≤–∞–µ–º —Ä–æ–¥–∏—Ç–µ–ª—å—Å–∫–∏–π –º–µ—Ç–æ–¥ –¥–ª—è –ü–ö–ú, —á—Ç–æ–±—ã –∏–∑–±–µ–∂–∞—Ç—å –∫–æ–Ω—Ñ–ª–∏–∫—Ç–æ–≤
            return

        # –î–ª—è –õ–ö–ú –∏ –¥—Ä—É–≥–∏—Ö –∫–Ω–æ–ø–æ–∫ –≤—ã–∑—ã–≤–∞–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—É—é –æ–±—Ä–∞–±–æ—Ç–∫—É
        super().mousePressEvent(event)

    def mouseMoveEvent(self, event):
        # –û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ–º –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏–µ –ü–ö–ú
        if self.right_dragging and event.buttons() & Qt.MouseButton.RightButton:
            self.right_drag_current_position = event.position().toPoint()

            # –ï—Å–ª–∏ –ø–µ—Ä–µ–º–µ—Å—Ç–∏–ª–∏—Å—å –¥–æ—Å—Ç–∞—Ç–æ—á–Ω–æ –¥–∞–ª–µ–∫–æ –æ—Ç –Ω–∞—á–∞–ª—å–Ω–æ–π —Ç–æ—á–∫–∏, –Ω–∞—á–∏–Ω–∞–µ–º –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏–µ
            if (
                    self.right_drag_start_position - self.right_drag_current_position).manhattanLength() > QtWidgets.QApplication.startDragDistance():
                self.startRightDrag()
                return

        # –î–ª—è –¥—Ä—É–≥–∏—Ö —Å–ª—É—á–∞–µ–≤ –≤—ã–∑—ã–≤–∞–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—É—é –æ–±—Ä–∞–±–æ—Ç–∫—É
        super().mouseMoveEvent(event)

    def mouseReleaseEvent(self, event):
        # –°–±—Ä–∞—Å—ã–≤–∞–µ–º —Å–æ—Å—Ç–æ—è–Ω–∏–µ –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏—è –ü–ö–ú
        if event.button() == Qt.MouseButton.RightButton:
            self.right_dragging = False
            self.right_drag_start_position = None
            self.right_drag_current_position = None
            self.right_drag_index = -1
            self.right_drag_pixmap = None
            return

        super().mouseReleaseEvent(event)

    def startRightDrag(self):
        if self.right_drag_index < 0:
            return

        drag = QtGui.QDrag(self)
        mime_data = QtCore.QMimeData()

        # –°–æ—Ö—Ä–∞–Ω—è–µ–º –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ –≤–∫–ª–∞–¥–∫–µ
        tab_data = QtCore.QByteArray()
        stream = QtCore.QDataStream(tab_data, QtCore.QIODevice.OpenModeFlag.WriteOnly)
        stream.writeQString(self.tabText(self.right_drag_index))
        stream.writeInt(self.right_drag_index)

        mime_data.setData("application/x-tabwidget", tab_data)
        drag.setMimeData(mime_data)

        # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –≤–∏–∑—É–∞–ª—å–Ω—É—é –æ–±—Ä–∞—Ç–Ω—É—é —Å–≤—è–∑—å - pixmap –≤–∫–ª–∞–¥–∫–∏
        if self.right_drag_pixmap:
            drag.setPixmap(self.right_drag_pixmap)
            # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —Ç–æ—á–∫—É "–∑–∞—Ö–≤–∞—Ç–∞" - –≥–¥–µ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å –∑–∞–∂–∞–ª –≤–∫–ª–∞–¥–∫—É
            drag.setHotSpot(self.right_drag_current_position - self.tabRect(self.right_drag_index).topLeft())

        # –ó–∞–ø—É—Å–∫–∞–µ–º –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏–µ
        result = drag.exec(Qt.DropAction.MoveAction)

        # –ï—Å–ª–∏ –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏–µ –∑–∞–≤–µ—Ä—à–∏–ª–æ—Å—å –≤–Ω–µ –æ–∫–Ω–∞, —Å–æ–∑–¥–∞–µ–º –Ω–æ–≤–æ–µ –æ–∫–Ω–æ
        if result == Qt.DropAction.IgnoreAction:
            self.createNewWindowWithTab()

    def createNewWindowWithTab(self):
        """–°–æ–∑–¥–∞–µ—Ç –Ω–æ–≤–æ–µ –æ–∫–Ω–æ —Å —Ç–µ–∫—É—â–µ–π –≤–∫–ª–∞–¥–∫–æ–π"""
        if self.right_drag_index < 0:
            return

        try:
            # –ü–æ–ª—É—á–∞–µ–º –¥–∞–Ω–Ω—ã–µ –æ –≤–∫–ª–∞–¥–∫–µ
            current_index = self.currentIndex()
            if current_index < 0:
                return

            widget = self.parent().widget(current_index)
            title = self.tabText(current_index)

            # –°–æ–∑–¥–∞–µ–º –Ω–æ–≤–æ–µ –æ–∫–Ω–æ
            new_window = MainWindow()

            # –£–¥–∞–ª—è–µ–º –≤–∫–ª–∞–¥–∫—É –∏–∑ —Ç–µ–∫—É—â–µ–≥–æ –æ–∫–Ω–∞
            self.parent().removeTab(current_index)

            # –î–æ–±–∞–≤–ª—è–µ–º –≤ –Ω–æ–≤–æ–µ –æ–∫–Ω–æ
            new_window.main_tab_widget.addTab(widget, title)
            new_window.main_tab_widget.setCurrentWidget(widget)

            # –ü–æ–∫–∞–∑—ã–≤–∞–µ–º –Ω–æ–≤–æ–µ –æ–∫–Ω–æ —Ä—è–¥–æ–º —Å –∫—É—Ä—Å–æ—Ä–æ–º
            global_pos = self.mapToGlobal(self.right_drag_current_position)
            new_window.move(global_pos.x() - 100, global_pos.y() - 50)
            new_window.show()

        except Exception as e:
            print(f"Error creating new window: {e}")

    def dragEnterEvent(self, event):
        if event.mimeData().hasFormat("application/x-tabwidget"):
            event.acceptProposedAction()

    def dragMoveEvent(self, event):
        if event.mimeData().hasFormat("application/x-tabwidget"):
            event.acceptProposedAction()

    def dropEvent(self, event):
        if event.mimeData().hasFormat("application/x-tabwidget"):
            tab_data = event.mimeData().data("application/x-tabwidget")
            stream = QtCore.QDataStream(tab_data, QtCore.QIODevice.OpenModeFlag.ReadOnly)
            title = stream.readQString()
            source_index = stream.readInt()

            # –ù–∞—Ö–æ–¥–∏–º –ø–æ–∑–∏—Ü–∏—é –¥–ª—è –≤—Å—Ç–∞–≤–∫–∏
            drop_pos = event.position().toPoint()
            insert_index = self.count()
            for i in range(self.count()):
                if self.tabRect(i).contains(drop_pos):
                    insert_index = i
                    break

            # –ï—Å–ª–∏ —ç—Ç–æ –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏–µ –∏–∑ –¥—Ä—É–≥–æ–≥–æ –æ–∫–Ω–∞
            if event.source() != self:
                # –ü–æ–ª—É—á–∞–µ–º –≤–∏–¥–∂–µ—Ç –∏–∑ –∏—Å—Ö–æ–¥–Ω–æ–≥–æ –æ–∫–Ω–∞
                source_widget = None
                source_parent = event.source().parent()
                if hasattr(source_parent, 'widget'):
                    source_widget = source_parent.widget(source_index)

                if source_widget:
                    # –î–æ–±–∞–≤–ª—è–µ–º –≤–∫–ª–∞–¥–∫—É –≤ —Ç–µ–∫—É—â–∏–π TabWidget
                    self.parent().insertTab(insert_index, source_widget, title)
                    self.parent().setCurrentIndex(insert_index)

                    # –£–¥–∞–ª—è–µ–º –≤–∫–ª–∞–¥–∫—É –∏–∑ –∏—Å—Ö–æ–¥–Ω–æ–≥–æ TabWidget
                    event.source().parent().removeTab(source_index)

            event.acceptProposedAction()


class DraggableTabWidget(QtWidgets.QTabWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setTabBar(DraggableTabBar(self))
        self.setDocumentMode(True)
        self.setElideMode(Qt.TextElideMode.ElideRight)
        self.setTabsClosable(True)
        self.setMovable(True)  # –í–∫–ª—é—á–∞–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω–æ–µ –ø–µ—Ä–µ–º–µ—â–µ–Ω–∏–µ –¥–ª—è –õ–ö–ú
        self.setAcceptDrops(True)

    def dragEnterEvent(self, event):
        if event.mimeData().hasFormat("application/x-tabwidget"):
            event.acceptProposedAction()

    def dragMoveEvent(self, event):
        if event.mimeData().hasFormat("application/x-tabwidget"):
            event.acceptProposedAction()

    def dropEvent(self, event):
        if event.mimeData().hasFormat("application/x-tabwidget"):
            tab_data = event.mimeData().data("application/x-tabwidget")
            stream = QtCore.QDataStream(tab_data, QtCore.QIODevice.OpenModeFlag.ReadOnly)
            title = stream.readQString()
            source_index = stream.readInt()

            # –ï—Å–ª–∏ –ø–µ—Ä–µ—Ç–∞—Å–∫–∏–≤–∞–Ω–∏–µ –∏–∑ –¥—Ä—É–≥–æ–≥–æ –æ–∫–Ω–∞
            if event.source() != self.tabBar():
                # –ü–æ–ª—É—á–∞–µ–º –≤–∏–¥–∂–µ—Ç –∏–∑ –∏—Å—Ö–æ–¥–Ω–æ–≥–æ –æ–∫–Ω–∞
                source_widget = None
                if hasattr(event.source(), 'parent') and hasattr(event.source().parent(), 'widget'):
                    source_widget = event.source().parent().widget(source_index)

                if source_widget:
                    # –î–æ–±–∞–≤–ª—è–µ–º –≤–∫–ª–∞–¥–∫—É
                    self.addTab(source_widget, title)
                    self.setCurrentWidget(source_widget)

                    # –£–¥–∞–ª—è–µ–º –≤–∫–ª–∞–¥–∫—É –∏–∑ –∏—Å—Ö–æ–¥–Ω–æ–≥–æ TabWidget
                    event.source().parent().removeTab(source_index)

            event.acceptProposedAction()


class TopBar(QWidget):
    workspace_changed = pyqtSignal(str)
    profile_requested = pyqtSignal()
    theme_toggle_requested = pyqtSignal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setFixedHeight(56)
        self.panel_menu_button = None
        self.panel_menu = None
        self.setup_ui()

    def setup_ui(self):
        layout = QHBoxLayout(self)
        layout.setContentsMargins(16, 6, 16, 6)
        layout.setSpacing(8)

        self.theme_button = QToolButton()
        self.theme_button.setText("üåì")
        self.theme_button.setToolTip("–ü–µ—Ä–µ–∫–ª—é—á–∏—Ç—å —Ç–µ–º—É")
        self.theme_button.setAutoRaise(True)
        self.theme_button.clicked.connect(self.theme_toggle_requested.emit)
        layout.addWidget(self.theme_button)

        layout.addStretch()

        self.workspace_combo = QComboBox()
        self.workspace_combo.setMinimumWidth(260)
        self.workspace_combo.addItems(["–ü–æ —É–º–æ–ª—á–∞–Ω–∏—é", "–†–∞–∑—Ä–∞–±–æ—Ç–∫–∞", "–ê–Ω–∞–ª–∏—Ç–∏–∫–∞", "–ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–µ"])
        self.workspace_combo.setToolTip("–í—ã–±–µ—Ä–∏—Ç–µ —Ä–∞–±–æ—á–µ–µ –ø—Ä–æ—Å—Ç—Ä–∞–Ω—Å—Ç–≤–æ")

        right_container = QHBoxLayout()
        right_container.setSpacing(6)

        self.profile_button = QToolButton()
        try:
            self.profile_button.setIcon(QIcon.fromTheme("user"))
        except Exception:
            pass
        self.profile_button.setText("–ü—Ä–æ—Ñ–∏–ª—å")
        self.profile_button.setToolTip("–û—Ç–∫—Ä—ã—Ç—å –ø—Ä–æ—Ñ–∏–ª—å")
        self.profile_button.setAutoRaise(True)

        self.panel_menu_button = QToolButton()
        try:
            self.panel_menu_button.setIcon(QIcon.fromTheme("view-list"))
        except Exception:
            pass
        self.panel_menu_button.setText("–ü–∞–Ω–µ–ª–∏")
        self.panel_menu_button.setToolTip("–ü–æ–∫–∞–∑–∞—Ç—å/—Å–∫—Ä—ã—Ç—å –ø–∞–Ω–µ–ª–∏")
        self.panel_menu_button.setAutoRaise(True)

        self.panel_menu = QMenu(self)

        def _show_panel_menu():
            btn = self.panel_menu_button
            pos = btn.mapToGlobal(QPoint(0, btn.height()))
            if self.panel_menu.actions():
                self.panel_menu.exec(pos)

        self.panel_menu_button.clicked.connect(_show_panel_menu)

        right_container.addWidget(self.profile_button)
        right_container.addWidget(self.panel_menu_button)

        center_widget = QWidget()
        center_layout = QHBoxLayout(center_widget)
        center_layout.setContentsMargins(0, 0, 0, 0)
        center_layout.addWidget(self.workspace_combo)
        center_layout.addStretch()

        layout.addWidget(center_widget)
        layout.addLayout(right_container)
        layout.addStretch()

        self.workspace_combo.currentTextChanged.connect(self.workspace_changed.emit)

    def set_panels_menu(self, actions):
        self.panel_menu.clear()
        for act in actions:
            self.panel_menu.addAction(act)

    def update_theme_button(self, theme):
        if theme == "dark":
            self.theme_button.setText("‚òÄÔ∏è")
            self.theme_button.setToolTip("–ü–µ—Ä–µ–∫–ª—é—á–∏—Ç—å –Ω–∞ —Å–≤–µ—Ç–ª—É—é —Ç–µ–º—É")
        else:
            self.theme_button.setText("üåô")
            self.theme_button.setToolTip("–ü–µ—Ä–µ–∫–ª—é—á–∏—Ç—å –Ω–∞ —Ç–µ–º–Ω—É—é —Ç–µ–º—É")


class MainWindow(QMainWindow):
    WORKSPACE_SETTINGS_KEY = "app/worrrkspace"

    def __init__(self):
        super().__init__()
        self.settings = QSettings("worrrkspace_company", "worrrkspace_app")

        self.theme_manager = ThemeManager(organization="worrrkspace_company", application="worrrkspace_app")
        self.current_theme = self.theme_manager.current_theme

        self.dock_widgets = {}
        self._panel_actions = {}
        self.setup_window()
        self.load_ui()
        self.setup_ui()
        self.apply_theme()
        self.setup_connections()

    def setup_window(self):
        self.setWindowTitle("WoRRRkspace")
        self.setMinimumSize(1000, 700)
        self.resize(1400, 900)
        screen = self.screen()
        if screen:
            geom = screen.geometry()
            center = geom.center()
            self.move(center - self.rect().center())

    def load_ui(self):
        current_dir = os.path.dirname(os.path.abspath(__file__))
        ui_file = os.path.join(current_dir, "MainWindow.ui")
        if os.path.exists(ui_file):
            try:
                uic.loadUi(ui_file, self)
                self.clear_builtin_styles()
                # –ó–∞–º–µ–Ω—è–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–π QTabWidget –Ω–∞ –Ω–∞—à DraggableTabWidget
                self.replace_tab_widget()
            except Exception as e:
                print(f"Error loading UI file: {e}")
                self.setup_fallback_ui()
        else:
            self.setup_fallback_ui()

    def replace_tab_widget(self):
        """–ó–∞–º–µ–Ω—è–µ—Ç —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–π QTabWidget –Ω–∞ –Ω–∞—à DraggableTabWidget"""
        if hasattr(self, 'main_tab_widget') and isinstance(self.main_tab_widget, QtWidgets.QTabWidget):
            old_tab_widget = self.main_tab_widget

            # –°–æ–∑–¥–∞–µ–º –Ω–æ–≤—ã–π –∫–∞—Å—Ç–æ–º–Ω—ã–π –≤–∏–¥–∂–µ—Ç –≤–∫–ª–∞–¥–æ–∫
            new_tab_widget = DraggableTabWidget()

            # –ü–µ—Ä–µ–Ω–æ—Å–∏–º –≤—Å–µ –≤–∫–ª–∞–¥–∫–∏ –∏–∑ —Å—Ç–∞—Ä–æ–≥–æ –≤–∏–¥–∂–µ—Ç–∞
            for i in range(old_tab_widget.count()):
                widget = old_tab_widget.widget(i)
                title = old_tab_widget.tabText(i)
                new_tab_widget.addTab(widget, title)

            # –ó–∞–º–µ–Ω—è–µ–º –≤ layout
            if hasattr(self, 'verticalLayout'):
                layout = self.verticalLayout
                index = layout.indexOf(old_tab_widget)
                layout.removeWidget(old_tab_widget)
                layout.insertWidget(index, new_tab_widget)
                self.main_tab_widget = new_tab_widget

                # –ü–æ–¥–∫–ª—é—á–∞–µ–º –æ–±—Ä–∞–±–æ—Ç—á–∏–∫ –∑–∞–∫—Ä—ã—Ç–∏—è –≤–∫–ª–∞–¥–æ–∫
                self.main_tab_widget.tabCloseRequested.connect(self.close_tab)

                old_tab_widget.deleteLater()

    def clear_builtin_styles(self):
        self.setStyleSheet("")
        if hasattr(self, 'centralwidget'):
            self.centralwidget.setStyleSheet("")
        if hasattr(self, 'main_tab_widget'):
            self.main_tab_widget.setStyleSheet("")
        if hasattr(self, 'welcome_tab'):
            self.welcome_tab.setStyleSheet("")
        if hasattr(self, 'settings_tab'):
            self.settings_tab.setStyleSheet("")
        for child in self.findChildren(QtWidgets.QWidget):
            child.setStyleSheet("")

    def setup_fallback_ui(self):
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        layout = QVBoxLayout(central_widget)
        layout.setContentsMargins(0, 0, 0, 0)
        self.main_tab_widget = DraggableTabWidget()  # –ò—Å–ø–æ–ª—å–∑—É–µ–º –Ω–∞—à –∫–∞—Å—Ç–æ–º–Ω—ã–π –≤–∏–¥–∂–µ—Ç
        self.main_tab_widget.setTabsClosable(True)
        layout.addWidget(self.main_tab_widget)
        self.status_bar = QStatusBar()
        self.setStatusBar(self.status_bar)

    def setup_ui(self):
        self.top_bar = TopBar()
        self.setMenuWidget(self.top_bar)
        self.top_bar.update_theme_button(self.current_theme)

        self.create_panels()

        if not hasattr(self, "status_bar") or self.statusBar() is None:
            self.status_bar = QStatusBar()
            self.setStatusBar(self.status_bar)

        saved_ws = self.settings.value(self.WORKSPACE_SETTINGS_KEY, "")
        if saved_ws:
            idx = self.top_bar.workspace_combo.findText(saved_ws)
            if idx >= 0:
                self.top_bar.workspace_combo.setCurrentIndex(idx)

        self.rebuild_panel_actions_menu()

    def create_panels(self):
        # –°–æ–∑–¥–∞–µ–º –ø–∞–Ω–µ–ª–∏ –∏–∑ –∏–º–ø–æ—Ä—Ç–∏—Ä–æ–≤–∞–Ω–Ω—ã—Ö –∫–ª–∞—Å—Å–æ–≤
        self.solution_explorer = SolutionExplorer(self)
        self.addDockWidget(Qt.DockWidgetArea.LeftDockWidgetArea, self.solution_explorer)
        self.dock_widgets["solution_explorer"] = self.solution_explorer

        self.tools_panel = ToolsPanel(self)
        self.addDockWidget(Qt.DockWidgetArea.LeftDockWidgetArea, self.tools_panel)
        self.dock_widgets["tools"] = self.tools_panel

        self.chat_panel = ChatPanel(self)
        self.addDockWidget(Qt.DockWidgetArea.RightDockWidgetArea, self.chat_panel)
        self.dock_widgets["chat"] = self.chat_panel

        self.logs_panel = LogsPanel(self)
        self.addDockWidget(Qt.DockWidgetArea.BottomDockWidgetArea, self.logs_panel)
        self.dock_widgets["logs"] = self.logs_panel

        try:
            self.tabifyDockWidget(self.solution_explorer, self.tools_panel)
            self.solution_explorer.raise_()
        except Exception:
            pass

        for key, dock in self.dock_widgets.items():
            dock.visibilityChanged.connect(self.rebuild_panel_actions_menu)

    def setup_connections(self):
        self.top_bar.workspace_changed.connect(self.on_workspace_changed)
        self.top_bar.profile_button.clicked.connect(self.open_profile)
        self.top_bar.theme_toggle_requested.connect(self.toggle_theme)

        if hasattr(self, "tools_panel"):
            tp = self.tools_panel
            tp.btn_table.clicked.connect(lambda: self._open_placeholder_tab("–¢–∞–±–ª–∏—Ü–∞"))
            tp.btn_note.clicked.connect(lambda: self._open_placeholder_tab("–ó–∞–º–µ—Ç–∫–∞"))
            # tp.btn_diagramm.clicked.connect(lambda: self._open_placeholder_tab("–î–∏–∞–≥—Ä–∞–º–º–∞"))
            tp.btn_chart.clicked.connect(lambda: self._open_placeholder_tab("–ì—Ä–∞—Ñ"))
            tp.btn_task.clicked.connect(lambda: self._open_placeholder_tab("–ó–∞–¥–∞—á–∞"))

        if hasattr(self, "main_tab_widget"):
            try:
                self.main_tab_widget.tabCloseRequested.connect(self.close_tab)
            except Exception:
                pass

    def _open_placeholder_tab(self, title: str):
        if not hasattr(self, "main_tab_widget"):
            return

        if title == "–ó–∞–º–µ—Ç–∫–∞":
            w = MarkdownNoteTab()
            w.parent = self
            self.main_tab_widget.addTab(w, "üìù " + title)
            self.main_tab_widget.setCurrentWidget(w)
        elif title == "–¢–∞–±–ª–∏—Ü–∞":
            w = TableEditorTab()
            w.parent = self
            self.main_tab_widget.addTab(w, "üìä " + title)
            self.main_tab_widget.setCurrentWidget(w)
        elif title == "–ì—Ä–∞—Ñ":
            w = GraphTab()
            w.parent = self
            self.main_tab_widget.addTab(w, "üï∏Ô∏è " + title)
            self.main_tab_widget.setCurrentWidget(w)
        elif title == "–ó–∞–¥–∞—á–∞":
            w = TaskTab()
            w.parent = self
            self.main_tab_widget.addTab(w, "‚úÖ " + title)
            self.main_tab_widget.setCurrentWidget(w)
        else:
            w = QWidget()
            l = QVBoxLayout(w)
            te = QTextEdit()
            te.setReadOnly(False)
            te.setPlainText(f"{title} ‚Äî —Å–æ–¥–µ—Ä–∂–∞–Ω–∏–µ (–∑–∞–≥–ª—É—à–∫–∞).")
            l.addWidget(te)
            self.main_tab_widget.addTab(w, title)
            self.main_tab_widget.setCurrentWidget(w)

    def rebuild_panel_actions_menu(self):
        actions = []
        self._panel_actions.clear()
        for key, dock in self.dock_widgets.items():
            title = dock.windowTitle() or key
            act = QAction(title, self)
            act.setCheckable(True)
            act.setChecked(dock.isVisible())

            def make_toggler(d=key):
                def toggle():
                    w = self.dock_widgets[d]
                    visible = not w.isVisible()
                    w.setVisible(visible)
                    if visible:
                        w.raise_()

                return toggle

            act.triggered.connect(make_toggler())
            actions.append(act)
            self._panel_actions[key] = act

        self.top_bar.set_panels_menu(actions)

    def close_tab(self, index):
        if self.main_tab_widget.count() > 1:
            self.main_tab_widget.removeTab(index)

    def on_workspace_changed(self, workspace_name: str):
        self.settings.setValue(self.WORKSPACE_SETTINGS_KEY, workspace_name)
        if hasattr(self, "status_bar"):
            self.status_bar.showMessage(f"–†–∞–±–æ—á–µ–µ –ø—Ä–æ—Å—Ç—Ä–∞–Ω—Å—Ç–≤–æ: {workspace_name}", 2500)

    def open_profile(self):
        dlg = ProfileDialog(self)
        dlg.exec()

    def apply_theme(self):
        app = QtWidgets.QApplication.instance()
        if not app:
            return

        setup_app_theme(app, self.current_theme)
        self.top_bar.update_theme_button(self.current_theme)
        self.force_style_update()

        if hasattr(self, "status_bar"):
            self.status_bar.showMessage(f"–¢–µ–º–∞: {self.current_theme}", 2000)

    def force_style_update(self):
        self.style().unpolish(self)
        self.style().polish(self)

        if self.centralWidget():
            self.centralWidget().style().unpolish(self.centralWidget())
            self.centralWidget().style().polish(self.centralWidget())

        self.top_bar.style().unpolish(self.top_bar)
        self.top_bar.style().polish(self.top_bar)

        for dock in self.dock_widgets.values():
            dock.style().unpolish(dock)
            dock.style().polish(dock)
            if dock.widget():
                dock.widget().style().unpolish(dock.widget())
                dock.widget().style().polish(dock.widget())

    def toggle_theme(self):
        self.current_theme = self.theme_manager.toggle_theme()
        self.apply_theme()


def main():
    app = QtWidgets.QApplication(sys.argv)
    setup_app_theme(app)
    w = MainWindow()
    w.show()
    sys.exit(app.exec())


if __name__ == "__main__":
    main()